OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -0,0 +1,328 @@
+			if (antiScrollPrevTopLeft.Y > wr.Viewport.TopLeft.Y)
+			{
+				// Get delta Y and limit to the max value
+				var tempRangeY = antiScrollPrevTopLeft.Y - wr.Viewport.TopLeft.Y;
+				var tempParticleCount = CalculateParticleCount(Game.Renderer.Resolution.Width, (int)tempRangeY);
+				if (ParticleList.Count + tempParticleCount > maxParticleCount)
+					tempParticleCount = maxParticleCount - ParticleList.Count;
+
+				SpawnParticles(tempParticleCount, (int)tempRangeY, 50);
+			}
+		}
+
+		void FadeOutParticleCount(WorldRenderer wr)
+		{
+			for (var i = 0; i < ParticleList.Count; i++)
+				if (ParticleList[i].PosY > (Game.Renderer.Resolution.Height - ParticleList[i].Gravity))
+					ParticleList.RemoveAt(i);
+		}
+
+		Particle XAxisSwing(Particle tempParticle)
+		{
+			// Direction turn
+			if (tempParticle.SwingOffset < -tempParticle.SwingAmplitude || tempParticle.SwingOffset > tempParticle.SwingAmplitude)
+				tempParticle.SwingDirection *= -1;
+
+			// Perform the X-Axis-Swing
+			tempParticle.SwingOffset += tempParticle.SwingDirection * tempParticle.SwingSpeed;
+			return tempParticle;
+		}
+
+		public void Tick(Actor self)
+		{
+			windTickCountdown--;
+		}
+
+		Particle WindLogic(Particle tempParticle)
+		{
+			if (!info.ChangingWind)
+				targetWindXOffset = info.WindStrength[0];
+			else if (windTickCountdown <= 0)
+			{
+				windTickCountdown = Game.CosmeticRandom.Next(info.WindTick[0], info.WindTick[1]);
+				if (Game.CosmeticRandom.Next(2) == 1 && currentWindIndex > 0)
+				{
+					currentWindIndex--;
+					targetWindXOffset = info.WindStrength[currentWindIndex];
+				}
+				else if (currentWindIndex < info.WindStrength.Length - 1)
+				{
+					currentWindIndex++;
+					targetWindXOffset = info.WindStrength[currentWindIndex];
+				}
+			}
+
+			// Fading the wind in little steps towards the TargetWindOffset
+			if (info.InstantWindChanges)
+				currentWindXOffset = targetWindXOffset;
+			else if (currentWindXOffset != targetWindXOffset)
+			{
+				if (currentWindXOffset > targetWindXOffset)
+					currentWindXOffset -= 0.00001f;
+				else if (currentWindXOffset < targetWindXOffset)
+					currentWindXOffset += 0.00001f;
+			}
+
+			return tempParticle;
+		}
+
+		Particle Movement(Particle tempParticle)
+		{
+			tempParticle.PosX += tempParticle.DirectionScatterX + tempParticle.SwingOffset + currentWindXOffset;
+			tempParticle.PosY += tempParticle.Gravity;
+			return tempParticle;
+		}
+
+		// AntiScroll keeps the particels in place when scrolling the viewport
+		Particle AntiScroll(Particle tempParticle, WorldRenderer wr)
+		{
+			if (antiScrollPrevTopLeft != wr.Viewport.TopLeft)
+			{
+				if (antiScrollPrevTopLeft.X != wr.Viewport.TopLeft.X)
+					tempParticle.PosX += antiScrollPrevTopLeft.X - wr.Viewport.TopLeft.X;
+
+				if (antiScrollPrevTopLeft.Y != wr.Viewport.TopLeft.Y)
+					tempParticle.PosY += antiScrollPrevTopLeft.Y - wr.Viewport.TopLeft.Y;
```

Comment:
It doesn't seem we need any of these three if statements here - the positions would just get updated by zero.

I'd suggest removing them for clarity.